packages/backend/src/core/BlockDownloader.ts

```diff
@@ -0,0 +1,218 @@
+import assert from 'assert'
+import { providers } from 'ethers'
+import { range } from 'lodash'
+
+import { json } from '../model'
+import {
+  BlockRecord,
+  BlockRepository,
+} from '../peripherals/database/BlockRepository'
+import { EthereumClient } from '../peripherals/ethereum/EthereumClient'
+import { BlockNumber, BlockRange } from '../peripherals/ethereum/types'
+import { createEventEmitter } from '../tools/EventEmitter'
+import { JobQueue } from '../tools/JobQueue'
+import { Logger } from '../tools/Logger'
+
+export interface KnownBlock {
+  number: number
+  hash: string
+}
+
+type State = { t: 'not-started' } | { t: 'working'; lastKnownBlock: KnownBlock }
+
+export class BlockDownloader {
+  private events = createEventEmitter<BlockDownloaderEvents>()
+  private state: State = { t: 'not-started' }
+  private jobQueue: JobQueue
+
+  constructor(
+    private ethereumClient: EthereumClient,
+    private blockRepository: BlockRepository,
+    private logger: Logger
+  ) {
+    this.logger = this.logger.for(this)
+    this.jobQueue = new JobQueue({ maxConcurrentJobs: 1 }, this.logger)
+  }
+
+  async start() {
+    const lastKnownBlock = await this.blockRepository.getLast()
+
+    this.state = { t: 'working', lastKnownBlock }
+
+    return this.startBackgroundWork()
+  }
+
+  getStatus() {
+    const status: json = { status: this.state.t }
+
+    if (this.state.t === 'working')
+      status.lastKnownBlock = { ...this.state.lastKnownBlock }
+
+    return status
+  }
+
+  getLastKnownBlock() {
+    if (this.state.t !== 'working') {
+      throw new Error('Not started')
+    }
+    return this.state.lastKnownBlock
+  }
+
+  onNewBlocks(handler: (blockRange: BlockRange) => void) {
+    this.events.on('newBlocks', handler)
```

Comment:
What about this being called immediately with some value if we know of a last block?

Reply:
I don't think this is a correct behaviour.

If we call the listener with _just the last block_, we're neither _hot_ nor _cold_.

Current implementation is _hot_ like an usual event emitter. The events are emitted regardless if somebody listens or not. 

Since we're now reacting on `block` events instead of scheduling work on interval, I moved to handling "sync on restart" behaviour in SyncScheduler. I still have to build the "empowered BlockRange" here, but you can notice I started emitting `{from, to}` instead of just `to: number`.

Take a look at the following.

```ts
    if (lastKnown.number > lastSynced) {
      this.scheduleSync({ from: lastSynced, to: lastKnown.number })
      lastSynced = lastKnown.number
    }

    this.blockDownloader.onNewBlocks((blockRange) => {
      lastSynced = blockRange.to
      this.scheduleSync(blockRange)
    })
```

Reply:
Makes sense. Let's continue that way

Reply:
Makes sense. Let's continue that way

Reply:
Alternatively, we coould make it entirely _cold_ and replay all unhandled events from the start of the downloader for every new listener. It would be "cleaner but trickier", so I'm not convinced if we wanna go in this direction.